tests/30rooms/01state.pl

```diff
@@ -237,3 +237,48 @@
          Future->done(1);
       });
    };
+
+
+sub set_test_state
+{
+   my ( $user, $room_id ) = @_;
+
+   matrix_put_room_state( $user, $room_id,
+         type      => "a.test.state.type",
+         state_key => "",
+         content   => { "a_key" => "a_value" },
+   )->then( sub {
+      matrix_initialsync_room( $user, $room_id, limit => 0 );
+   })->then( sub {
+      my ( $body ) = @_;
+
+      my %state_by_type = partition_by { $_->{type} } @{ $body->{state} };
+
+      my $event_id = $state_by_type{"a.test.state.type"}[0]{event_id};
+
+      Future->done( $event_id );
+   });
+}
+
+
+test "Setting state twice is idempotent",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      my ( $event_id_1, $event_id_2 );
+
+      set_test_state( $user, $room_id )->then( sub {
+         ( $event_id_1 ) = @_;
+
+         set_test_state( $user, $room_id );
+      })->then( sub {
+         ( $event_id_2 ) = @_;
+
+         $event_id_1 eq $event_id_2 or die "Did not expect a new event when"
+             . " uploading the same state event a second time";
```

Comment:
You can avoid that awkward-to-read string concat in the message if you

    $event_id_1 eq $event_id_2 or
       die "Did not expect ....";

As per STYLE-GUIDE it doesn't overly matter if simply the end of the message string spills over 80 chars; a reader with a terminal this narrow can infer the end of string closing quote and `;`.